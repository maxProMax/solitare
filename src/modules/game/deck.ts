import { makeObservable, action, observable } from "mobx";
import random from "lodash/random";
import { Suit, Type, Card } from "./card";

export class Deck {
  private _cards: Card[] = [];

  constructor() {
    const tempCards: (Card | null)[] = [];
    const cards: Card[] = [];

    for (const [, type] of Object.entries(Type)) {
      for (const [, suit] of Object.entries(Suit)) {
        tempCards.push(new Card({ suit, type }));
      }
    }

    while (tempCards.length !== cards.length) {
      const i = this.getIdx();

      if (tempCards.length === cards.length) {
        break;
      }

      if (tempCards[i]) {
        const { suit, type } = tempCards[i];
        cards.push(new Card({ suit, type }));
        tempCards[i] = null;
        continue;
      }

      if (tempCards.length - cards.length === 1) {
        const card = tempCards.find((c) => c);

        if (card) {
          const { suit, type } = card;
          cards.push(new Card({ suit, type }));

          break;
        }
      }
    }

    this.addToCards(...cards);

    makeObservable<Deck, "_cards">(this, {
      _cards: observable,
      addToCards: action,
      getCards: action,
      removeAllCards: action,
    });
  }

  get cards() {
    return this._cards;
  }

  addToCards(...cards: Card[]) {
    this._cards.push(...cards);
  }

  getIdx() {
    return random(0, 51);
  }

  /** Get cards by fixed amount */
  getCards(amount: number): Card[] {
    return this._cards.splice(0, amount);
  }

  removeAllCards() {
    this._cards = [];
  }

  getAllCards() {
    const cards = this._cards;

    this.removeAllCards();

    return cards;
  }
}
