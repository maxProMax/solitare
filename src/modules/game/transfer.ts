import { Card } from "./card";

/**
 * Temporary state between transferring cards
 */
// fromInstance?.removeTransferredCards();
interface CardsHolder {
  removeTransferredCards: () => void;
}

export class Transfer {
  private _cards: Card[] = [];
  private _fromInstance?: CardsHolder;

  set cards(cards: Card[]) {
    this._cards = cards;
  }

  get cards() {
    return this._cards;
  }

  get fromInstance() {
    return this._fromInstance;
  }

  addCards(fromInstance: CardsHolder, cards: Card[]) {
    this._cards = cards;
    this._fromInstance = fromInstance;
  }

  getCardAndRestTransfer() {
    const cards = this._cards;

    this._cards = [];

    this._fromInstance = undefined;

    return cards;
  }
}
