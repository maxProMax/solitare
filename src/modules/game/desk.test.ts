import { Deck } from "./deck";
import { Transfer } from "./transfer";

describe("Desk", () => {
  beforeEach(() => {
    const indexes = [
      51, 51, 51, 0, 50, 1, 49, 2, 48, 3, 47, 4, 46, 5, 45, 6, 44, 7, 43, 8, 42,
      9, 41, 10, 40, 11, 39, 12, 38, 13, 37, 14, 36, 15, 35, 16, 34, 17, 33, 18,
      32, 19, 31, 20, 30, 21, 29, 30, 30, 30, 22, 28, 23, 27, 24, 26, 25,
    ];
    jest
      .spyOn(Deck.prototype, "getIdx")
      .mockImplementation(() => indexes.shift() as number);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("check desk", () => {
    const deck = new Deck(new Transfer());

    expect(deck.cards).toMatchSnapshot();
    expect(2).toBe(2);
  });

  it("get 5 cards", () => {
    const deck = new Deck(new Transfer());

    deck.getCards(5);

    expect(deck.cards).toHaveLength(47);
  });
});
