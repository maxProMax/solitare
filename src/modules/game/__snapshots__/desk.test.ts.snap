// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desk check desk 1`] = `
[
  Card {
    "id": "♠️_10",
    "isOpen": false,
    "suit": "♠️",
    "type": "10",
  },
  Card {
    "id": "♣️_K",
    "isOpen": false,
    "suit": "♣️",
    "type": "K",
  },
  Card {
    "id": "♥️_10",
    "isOpen": false,
    "suit": "♥️",
    "type": "10",
  },
  Card {
    "id": "♦️_K",
    "isOpen": false,
    "suit": "♦️",
    "type": "K",
  },
  Card {
    "id": "♦️_10",
    "isOpen": false,
    "suit": "♦️",
    "type": "10",
  },
  Card {
    "id": "♥️_K",
    "isOpen": false,
    "suit": "♥️",
    "type": "K",
  },
  Card {
    "id": "♣️_10",
    "isOpen": false,
    "suit": "♣️",
    "type": "10",
  },
  Card {
    "id": "♠️_K",
    "isOpen": false,
    "suit": "♠️",
    "type": "K",
  },
  Card {
    "id": "♠️_9",
    "isOpen": false,
    "suit": "♠️",
    "type": "9",
  },
  Card {
    "id": "♣️_Q",
    "isOpen": false,
    "suit": "♣️",
    "type": "Q",
  },
  Card {
    "id": "♥️_9",
    "isOpen": false,
    "suit": "♥️",
    "type": "9",
  },
  Card {
    "id": "♦️_Q",
    "isOpen": false,
    "suit": "♦️",
    "type": "Q",
  },
  Card {
    "id": "♦️_9",
    "isOpen": false,
    "suit": "♦️",
    "type": "9",
  },
  Card {
    "id": "♥️_Q",
    "isOpen": false,
    "suit": "♥️",
    "type": "Q",
  },
  Card {
    "id": "♣️_9",
    "isOpen": false,
    "suit": "♣️",
    "type": "9",
  },
  Card {
    "id": "♠️_Q",
    "isOpen": false,
    "suit": "♠️",
    "type": "Q",
  },
  Card {
    "id": "♠️_8",
    "isOpen": false,
    "suit": "♠️",
    "type": "8",
  },
  Card {
    "id": "♣️_J",
    "isOpen": false,
    "suit": "♣️",
    "type": "J",
  },
  Card {
    "id": "♥️_8",
    "isOpen": false,
    "suit": "♥️",
    "type": "8",
  },
  Card {
    "id": "♦️_J",
    "isOpen": false,
    "suit": "♦️",
    "type": "J",
  },
  Card {
    "id": "♦️_8",
    "isOpen": false,
    "suit": "♦️",
    "type": "8",
  },
  Card {
    "id": "♥️_J",
    "isOpen": false,
    "suit": "♥️",
    "type": "J",
  },
  Card {
    "id": "♣️_8",
    "isOpen": false,
    "suit": "♣️",
    "type": "8",
  },
  Card {
    "id": "♠️_J",
    "isOpen": false,
    "suit": "♠️",
    "type": "J",
  },
  Card {
    "id": "♠️_7",
    "isOpen": false,
    "suit": "♠️",
    "type": "7",
  },
  Card {
    "id": "♣️_A",
    "isOpen": false,
    "suit": "♣️",
    "type": "A",
  },
  Card {
    "id": "♥️_7",
    "isOpen": false,
    "suit": "♥️",
    "type": "7",
  },
  Card {
    "id": "♦️_A",
    "isOpen": false,
    "suit": "♦️",
    "type": "A",
  },
  Card {
    "id": "♦️_7",
    "isOpen": false,
    "suit": "♦️",
    "type": "7",
  },
  Card {
    "id": "♥️_A",
    "isOpen": false,
    "suit": "♥️",
    "type": "A",
  },
  Card {
    "id": "♣️_7",
    "isOpen": false,
    "suit": "♣️",
    "type": "7",
  },
  Card {
    "id": "♠️_A",
    "isOpen": false,
    "suit": "♠️",
    "type": "A",
  },
  Card {
    "id": "♠️_6",
    "isOpen": false,
    "suit": "♠️",
    "type": "6",
  },
  Card {
    "id": "♣️_2",
    "isOpen": false,
    "suit": "♣️",
    "type": "2",
  },
  Card {
    "id": "♥️_6",
    "isOpen": false,
    "suit": "♥️",
    "type": "6",
  },
  Card {
    "id": "♦️_2",
    "isOpen": false,
    "suit": "♦️",
    "type": "2",
  },
  Card {
    "id": "♦️_6",
    "isOpen": false,
    "suit": "♦️",
    "type": "6",
  },
  Card {
    "id": "♥️_2",
    "isOpen": false,
    "suit": "♥️",
    "type": "2",
  },
  Card {
    "id": "♣️_6",
    "isOpen": false,
    "suit": "♣️",
    "type": "6",
  },
  Card {
    "id": "♠️_2",
    "isOpen": false,
    "suit": "♠️",
    "type": "2",
  },
  Card {
    "id": "♠️_5",
    "isOpen": false,
    "suit": "♠️",
    "type": "5",
  },
  Card {
    "id": "♣️_3",
    "isOpen": false,
    "suit": "♣️",
    "type": "3",
  },
  Card {
    "id": "♥️_5",
    "isOpen": false,
    "suit": "♥️",
    "type": "5",
  },
  Card {
    "id": "♦️_3",
    "isOpen": false,
    "suit": "♦️",
    "type": "3",
  },
  Card {
    "id": "♦️_5",
    "isOpen": false,
    "suit": "♦️",
    "type": "5",
  },
  Card {
    "id": "♥️_3",
    "isOpen": false,
    "suit": "♥️",
    "type": "3",
  },
  Card {
    "id": "♣️_5",
    "isOpen": false,
    "suit": "♣️",
    "type": "5",
  },
  Card {
    "id": "♠️_3",
    "isOpen": false,
    "suit": "♠️",
    "type": "3",
  },
  Card {
    "id": "♠️_4",
    "isOpen": false,
    "suit": "♠️",
    "type": "4",
  },
  Card {
    "id": "♣️_4",
    "isOpen": false,
    "suit": "♣️",
    "type": "4",
  },
  Card {
    "id": "♥️_4",
    "isOpen": false,
    "suit": "♥️",
    "type": "4",
  },
  Card {
    "id": "♦️_4",
    "isOpen": false,
    "suit": "♦️",
    "type": "4",
  },
]
`;
